data/shaders/opengl/rayleigh.glsl

```diff
@@ -0,0 +1,237 @@
+float height(const in vec3 orig, const in vec3 center)
+{
+	vec3 r = orig - center;
+	float height = sqrt(dot(r, r)) - geosphereRadius;
+
+	return height;
+}
+
+void scatter(out vec2 density, const in vec3 orig, const in vec3 center)
+{
+	float height = height(orig, center);
+
+	density = -height / baricStep;
+
+	// earth atmospheric density: 1.225 kg/m^3, divided by 1e5
+	// 1/1.225e-5 = 81632.65306
+	float earthDensities = geosphereAtmosFogDensity * 81632.65306f;
+	density /= earthDensities;
+}
+
+void findClosestHeight(out float h, out float t, const in vec3 orig, const in vec3 dir, const in vec3 center)
+{
+	vec3 radiusVector = center - orig;
+	vec3 tangent = dot(dir, radiusVector) * dir;
+	vec3 normal = radiusVector - tangent;
+	h = sqrt(dot(normal, normal));
+	t = dot(tangent, dir);
+}
+
+// error function
+float erf(const in float x)
+{
+	float a = 0.14001228904;
+	float four_over_pi = 1.27323954;
+
+	float x2 = x*x;
+	float r = -x2 * (four_over_pi + a * x2) / (1 + a * x2);
+	if (x > 0)
+		return  sqrt(1-exp(r)) * 0.5 + 0.5;
+	else
+		return -sqrt(1-exp(r)) * 0.5 + 0.5;
+}
+
+// predict out-scattering density based on coefficients
+float predictDensityOut(const in float atmosphereHeight, const in float height, const in float k, const in float b)
+{
+	if (height > atmosphereHeight)
+		return 0.f;
+
+	// earth atmospheric density: 1.225 kg/m^3, divided by 1e5
+	// 1/1.225e-5 = 81632.65306
+	float earthDensities = geosphereAtmosFogDensity * 81632.65306f;
+
+	if (height < 0)
+		return k / earthDensities;
+
+	return k * exp(-height * b) / earthDensities;
+}
+
+// predict in-scattering rate: from 0 to 1
+float predictDensityIn(const in float radius, const in float atmosphereHeight, const in float height, const in float c, const in float t)
+{
+	float minHeight = radius + height;
+	float h = sqrt(minHeight*minHeight + t*t); // height for our position
+
+	if (h > radius + atmosphereHeight) {
+		if (t > 0)
+			return 1.f; // no in-scattering, looking towards atmosphere
+		else
+			return 0.f; // looking from atmosphere
+	} else {
+		return erf(c * t); // erf is monotonic ascending
+	}
+}
+
+float predictDensityInOut(const in vec3 sample, const in vec3 dir, const in vec3 center, const in float radius, const in float atmosphereHeight, const in vec3 coefficients)
+{
+    float h, t;
+    findClosestHeight(h, t, sample, dir, center);
+    h -= radius;
+
+    float opticalDepth = 0.f;
+    // find out-scattering density
+    opticalDepth = predictDensityOut(atmosphereHeight, h, coefficients.x, coefficients.y);
+
+    // apply in-scattering filter
+    opticalDepth *= predictDensityIn(radius, atmosphereHeight, h, coefficients.z, t);
+    return opticalDepth;
+}
+
+void skipRay(inout vec2 opticalDepth, const in vec3 dir, const in vec2 boundaries, const in vec3 center)
+{
+	if (boundaries.y == boundaries.x)
+		return;
+
+	int numSamples = 16;
```

Comment:
I'd recommend lowering the number of samples used by the skip ray as much as possible without compromising the quality of the resulting scattering. In my testing a high sample count here did not significantly contribute to the visual quality of atmospheres.

Reply:
I'll leave 8